### ISSUE TITLE
Rule Change: Add option to no-param-reassign to allow reassignment to params with certain names

### ISSUE THREAD

**Issue by openjck** [eslint/eslint#15244]
### What rule do you want to change?

no-param-reassign

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
function setCount(state, action) {
  state = action.payload;
}
```

### What does the rule currently do for this code?

If `no-param-reassign` is set to `error`, the following error is raised.

```
no-param-reassign: Assignment to function parameter 'state'.
```

### What will the rule do after it's changed?

If `no-param-reassign` were set to `error` and the rule were configured such that reassignments to `state` were allowed, ESLint would not raise an error.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I would like to request an option for `no-param-reassign` which allows reassignment of params with certain names. For example, something like this, which would cause my example code above to pass:

```
  "no-param-reassign": ["error", { ignoreParamsNamed: ["state"] } ]
```

I know that there is an option called `ignorePropertyModificationsFor` which allows properties of params with certain names to be modified, but this request is different. I'm requesting an option which would allow the param itself to be reassigned if it has a certain name.

This would be useful for people who want to avoid param reassignment, but who want to allow it when using something like [Immer](https://github.com/immerjs/immer), which encourages reassignment and, by convention, often involves params with predictable names, like "state", "draft", and "draftState".

See [this comment](https://github.com/immerjs/immer/issues/189#issuecomment-844818586), which provides an example of someone who would benefit from this option.

Thank you! 🙂

**Comment by mdjermanovic** [comment]
Hi @openjck, thanks for the issue!

> ```js
> function setCount(state, action) {
>   state = action.payload;
> }
> ```

This looks like a possible error. The function reassigns a local `state` variable, so at the end it does nothing.

> This would be useful for people who want to avoid param reassignment, but who want to allow it when using something like [Immer](https://github.com/immerjs/immer), which encourages reassignment

I'm not sure why this would be encouraged, can you point to the documentation where it recommends param reassignment?

> See [this comment](https://github.com/immerjs/immer/issues/189#issuecomment-844818586), which provides an example of someone who would benefit from this option.

That looks like a bug, too, as `state = someNewStateValue` has no effect.

**Comment by openjck** [comment]
Thanks for the quick response!

> This looks like a possible error. The function reassigns a local `state` variable, so at the end it does nothing.

The example function doesn't do anything, but could be rewritten to return `state`, for example.

> I'm not sure why this would be encouraged, can you point to the documentation where it recommends param reassignment?
> That looks like a bug, too, as state = someNewStateValue has no effect.

You're right, that is a bug. Immer allows authors to write code as if they are mutating an object without actually mutating it ([see this documentation](https://immerjs.github.io/immer/#with-immer)), but I only now realized that this behavior [does not extend to reassignment](https://immerjs.github.io/immer/pitfalls/#dont-reassign-the-recipe-argument).

In that case, I think my request is based on a flawed assumption: in fact, developers should *not* be reassigning `state` in Redux Toolkit, which uses Immer under the hood. If you don't think this proposed option would be useful for any other purpose, feel free to close. Apologies for my confusion.

**Comment by mdjermanovic** [comment]
> in fact, developers should _not_ be reassigning `state` in Redux Toolkit, which uses Immer under the hood.

I agree with closing, as it looks like this option would actually be dangerous because it would hide possible errors.
